extensions/map-layers-auth/src/ArcGis/ArcGisAccessClient.ts

```diff
@@ -0,0 +1,365 @@
+
+  /* Optional expiration after which the token will expire. Defined in minutes with a maximum of two weeks (20160 minutes)*/
+  tokenExpiration?: number;
+
+  /* Application client Ids */
+  clientIds: ArcGisOAuthClientIds;
+}
+
+/** @beta */
+export class ArcGisAccessClient implements MapLayerAccessClient {
+  public readonly onOAuthProcessEnd = new BeEvent();
+  private _redirectUri: string | undefined;
+  private _expiration: number | undefined;
+  private _clientIds: ArcGisOAuthClientIds | undefined;
+
+  public constructor() {
+  }
+
+  public initialize(oAuthConfig?: ArcGisOAuthConfig): boolean {
+    if (oAuthConfig) {
+      this._redirectUri = oAuthConfig.redirectUri;
+      this._expiration = oAuthConfig.tokenExpiration;
+      this._clientIds = oAuthConfig.clientIds;
+
+      this.initOauthCallbackFunction();
+    }
+    return true;
+  }
+
+  private initOauthCallbackFunction() {
+    (window as any).arcGisOAuth2Callback = (redirectLocation?: Location) => {
+      let eventSuccess = false;
+      let stateData;
+
+      if (redirectLocation && redirectLocation.hash.length > 0) {
+        const locationHash = redirectLocation.hash;
+        const hashParams = new URLSearchParams(locationHash.substring(1));
+        const token = hashParams.get("access_token") ?? undefined;
+        const expiresInStr = hashParams.get("expires_in") ?? undefined;
+        const userName = hashParams.get("username") ?? undefined;
+        const ssl = hashParams.get("ssl") === "true";
+        const stateStr = hashParams.get("state") ?? undefined;
+        const persist = hashParams.get("persist") === "true";
+        if (token !== undefined && expiresInStr !== undefined && userName !== undefined && ssl !== undefined && stateStr !== undefined) {
+          let endpointOrigin;
+          try {
+            const state = JSON.parse(stateStr);
+            stateData = state?.customData;
+            endpointOrigin = state?.endpointOrigin;
+
+          } catch {
+          }
+          const expiresIn = Number(expiresInStr);
+          const expiresAt = (expiresIn * 1000) + (+new Date());   // Converts the token expiration delay (seconds) into a timestamp (UNIX time)
+          if (endpointOrigin !== undefined) {
+            ArcGisTokenManager.setOAuth2Token(endpointOrigin, { token, expiresAt, ssl, userName, persist });
+            eventSuccess = true;
+          }
+
+        }
+      }
+      this.onOAuthProcessEnd.raiseEvent(eventSuccess, stateData);
+    };
+  }
+
+  public unInitialize() {
+    this._redirectUri = undefined;
+    this._expiration = undefined;
+    (window as any).arcGisOAuth2Callback = undefined;
+  }
+
+  public async getAccessToken(params: MapLayerAccessTokenParams): Promise<MapLayerAccessToken | undefined> {
+    // First lookup Oauth2 tokens, otherwise check try "legacy tokens" if credentials were provided
+    try {
+      const oauth2Token = await this.getOAuthTokenForMapLayerUrl(params.mapLayerUrl.toString());
```

Comment:
Where is the OAuth request being sent to retrieve this access token?
I trace the execution through [`ArcGisAccessClient.getOAuthTokenForMapLayerUrl`](https://github.com/iTwin/itwinjs-core/blob/887ccb7db57aba67150253cbf89ee9e9951bdb3b/extensions/map-layers-auth/src/ArcGis/ArcGisAccessClient.ts#L217) to [`ArcGisTokenManager.getOAuth2Token`](https://github.com/iTwin/itwinjs-core/blob/887ccb7db57aba67150253cbf89ee9e9951bdb3b/extensions/map-layers-auth/src/ArcGis/ArcGisTokenManager.ts#L60), but it terminates at [`ArcGisTokenManager.loadFromBrowserStorage()`](https://github.com/iTwin/itwinjs-core/blob/887ccb7db57aba67150253cbf89ee9e9951bdb3b/extensions/map-layers-auth/src/ArcGis/ArcGisTokenManager.ts#L100) without any path that triggers an OIDC popup, iframe, or redirect.

How does the browser storage that's being checked here get populated via OAuth?

I notice the callback function [here ](https://github.com/iTwin/itwinjs-core/blob/887ccb7db57aba67150253cbf89ee9e9951bdb3b/extensions/map-layers-auth/src/ArcGis/ArcGisAccessClient.ts#L75) that's capable of setting the token, but what code navigates me there through the auth provider?

Reply:
Here is how it works:

1.  `MapUrlDialog` calls `MapLayerSource.validateSource`, and receives a `MapLayerSourceStatus.RequireAuth` status.
2. MapLayerAccessClient.getTokenServiceEndPoint is being call to get a `MapLayerTokenEndpoint` object.
3. If a token endpoint was provided, it uses `MapLayerTokenEndpoint.getLoginUrl` to get the URL that will be used in the popup (see `useCrossOriginPopup)` to trigger the OAuth process. (If no endpoint if provided, we fallback to a legacy process where credentials are asked directly in `MapUrlDialog`)
4. When the OAuth process completes, the page will be redirected to the the `ArcGisOauthRedirect.tsx` page, which will call `arcGisOAuth2Callback` function defined by `ArcGisAccessClient.initialize`
5. As you already discovered,  the `arcGisOAuth2Callback` function is responsible of calling `ArcGisTokenManager.setOAuth2Token` which will populate `ArcGisTokenManager._oauth2Cache` and also call `ArcGisTokenManager.saveToBrowserStorage`.
6. Note that we only store generated token in the session storage, and not local browser storage, no matter if user has checked the 'Keep me signed in' toggle.

Reply:
@evelynpreslar-bentley Are we good here?